File: go/2023/day02a.go
//go:build ignore

package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

func main() {
	fmt.Println(solve())
}

func solve() int {
	var total int

	scanner := bufio.NewScanner(bufio.NewReader(os.Stdin))
	for scanner.Scan() {
		// Game 1: 3 blue, 4 red; 1 red, 2 green, 6 blue; 2 green
		_gameId, games, _ := strings.Cut(scanner.Text(), ": ")
		gameId, _ := strconv.Atoi(strings.Replace(_gameId, "Game ", "", 1))

		if validate(games) {
			total += gameId
		}
	}

	return total
}

// games = "3 blue, 4 red; 1 red, 2 green, 6 blue; 2 green"
func validate(games string) bool {
	validCubes := map[string]int{
		"red":   12,
		"green": 13,
		"blue":  14,
	}

	for _, game := range strings.Split(games, "; ") {
		for _, cubes := range strings.Split(game, ", ") {
			_amount, color, _ := strings.Cut(cubes, " ")
			amount, _ := strconv.Atoi(_amount)

			if amount > validCubes[color] {
				return false
			}
		}
	}

	return true
}

File: py/2023/day02a.py
# games = "3 blue, 4 red; 1 red, 2 green, 6 blue; 2 green"
def validate(games: str) -> bool:
    valid_cubes = {"red": 12, "green": 13, "blue": 14}

    for game in games.split("; "):
        for cubes in game.split(", "):
            amount, color = cubes.split(" ")
            if int(amount) > valid_cubes[color]:
                return False

    return True


total = 0

for line in open(0):
    game_id, games = line.strip("\n").replace("Game ", "").split(": ")

    if validate(games):
        total += int(game_id)

print(total)
